### PR TITLE
Combobox mult object

### PR DESCRIPTION
**Pull request by anitavincent**
#### What is the purpose of this pull request?
Adding custom rendering to comboboxmultiple and add docs for custom rendering.

# Breaking changes:

 - **useCombobox returns matches and setMatches now, these should be used to control the rendered list, not "list" anymore.**

- for useComboboxMultiple state **removeSelected was renamed to unselect**

- for useComboboxMultiple state **selected and setSelected have been removed**, to manage the selected items you should use `selectedItems`, and `setSelectedItems`.

#### How should this be manually tested?

Check out the built docs, the built storybook, check the code examples and see if they make sense.

#### Screenshots or example usage
image

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Quality improvement (tests or refactors)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (small fix or feature that doesn't impact functionalities)
- [x] Requires change to documentation, which has been updated accordingly

### CODE REVIEW

File: packages/admin-ui-docs/docs/guidelines/experimental/combobox.mdx [hunk 1]
```diff
@@ -254,21 +411,27 @@ Features and fixes scheduled for this component:
 | Name      | Type       | Description                  | Required | Default |
 | --------- | ---------- | ---------------------------- | -------- | ------- |
 | timeoutMs | `number`   | Timeout of the deferredValue | 🚫       | -       |
-| list      | `string[]` | List of items                | 🚫       | -       |
+| list      | `ItemType[]` | List of items                | 🚫       | -       |
+| getOptionValue | `(item: ItemType) => string` | Function that transforms item shape into a string value, no need to use it on string[] lists  | 🚫       | by default it returns the item value, stringified if it is not string|
```
Comment (kevinch) [inline packages/admin-ui-docs/docs/guidelines/experimental/combobox.mdx:415]:
```suggestion
| getOptionValue | `(item: ItemType) => string` | Function that transforms item shape into a string value, no need to use it on string[] lists  | 🚫       | by default it returns the item value, stringified if it is not a string|
```

Comment (kevinch) [reply packages/admin-ui-docs/docs/guidelines/experimental/combobox.mdx:415]: Not sure on this one. What do you think?

File: packages/admin-ui-docs/docs/guidelines/experimental/combobox.mdx [hunk 2]
```diff
@@ -277,24 +440,29 @@ It extends the `useComboboxState` props, and:
...
 | ------------------- | ---------- | ------------------------------------------------ | -------- | ------- |
 | shouldClearOnSelect | `boolean`  | If should clear the text input on select an item | 🚫       | -       |
-| defaultSelected     | `string[]` | List of items that are selected by default       | 🚫       | -       |
+| defaultSelected     | `ItemType[]` | List of items that are selected by default       | 🚫       | -       |
+| renderTag          | `(item: ItemType) => ReactNode` | Function that controls what's rendered inside tag       | 🚫       | copies renderOption prop's value |
 
 #### ComboboxMultipleState
 
-The hooks return the `ComboboxMultipleState`, that extends `ComboboxState` and:
+The hooks returns `ComboboxMultipleState`, which extends `ComboboxState` and additionally has:
```
Comment (kevinch) [inline packages/admin-ui-docs/docs/guidelines/experimental/combobox.mdx:448]:
```suggestion
The hook returns `ComboboxMultipleState`, which extends `ComboboxState` and additionally has:
```

File: packages/admin-ui/src/combobox/combobox-item.ts [hunk 3]
```diff
@@ -1,21 +1,18 @@
...
 
-export const ComboboxItem = createComponent<'div', ComboboxItemOptions>(
+export const ComboboxItem = createComponent<typeof AriakitComboboxItem>(
   (props) => {
-    // weird, the input of the hook should be exactly ComboboxItemOptions
-    const htmlProps = useComboboxItem(props as any) as Omit<
-      HTMLProps<'div'>,
-      never
-    >
+    // TODO: Check the type error that forbids onClick param
```
Comment (kevinch) [inline packages/admin-ui/src/combobox/combobox-item.ts]: Is this an issue on our side?

File: packages/admin-ui/src/combobox/combobox-multiple-item.tsx [hunk 4]
```diff
@@ -17,14 +17,20 @@ export function ComboboxMultipleItem(props: ComboboxMultipleItemProps) {
     className: cn(style.itemMultiple),
   }
 
+  // TODO: Check the type error that forbids onClick param
```
Comment (kevinch) [inline packages/admin-ui/src/combobox/combobox-multiple-item.tsx:20]: Idem.

File: packages/admin-ui/src/combobox/combobox-multiple.state.ts [hunk 5]
```diff
@@ -18,37 +38,74 @@ export function useComboboxMultipleState(
...
+      )
+    })
+  }
+
   useEffect(() => {
     if (shouldClearOnSelect) combobox.setValue('')
   }, [checkbox.value, shouldClearOnSelect, combobox.setValue])
 
+  useEffect(() => {
+    if (shouldClearOnSelect) combobox.setValue('')
```
Comment (kevinch) [inline packages/admin-ui/src/combobox/combobox-multiple.state.ts:80]: Same as above? (line 76)

File: packages/admin-ui/src/combobox/combobox-item.ts (hunk 3 above)
Comment (anitavincent) [reply packages/admin-ui/src/combobox/combobox-item.ts]: We're not sure yet

File: packages/admin-ui-docs/docs/guidelines/experimental/combobox.mdx (hunk 1 above)
Comment (anitavincent) [reply packages/admin-ui-docs/docs/guidelines/experimental/combobox.mdx:415]: I see "not string" like not of the type string. Like you would say "not int"

File: packages/admin-ui/src/combobox/combobox-multiple.state.ts (hunk 5 above)
Comment (anitavincent) [reply packages/admin-ui/src/combobox/combobox-multiple.state.ts:80]: Don't know how I ended up with this here 😂

File: packages/admin-ui/src/combobox/checkbox-object-state.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    isSelected,
+    value: selectedItems,
+    setValue: setSelectedItems,
+  }
+}
+
+interface CheckboxObjectStateReturn<T> {
+  onChange: (item: T) => void
+  onAdd: (item: T) => void
+  onRemove: (item: T) => void
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/checkbox-object-state.ts:42]:
```suggestion
  toggle: (item: T) => void
  select: (item: T) => void
  unselect: (item: T) => void
```

File: packages/admin-ui/src/combobox/checkbox-object-state.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+  return {
+    onChange,
+    onAdd,
+    onRemove,
+    isSelected,
+    value: selectedItems,
+    setValue: setSelectedItems,
+  }
+}
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/checkbox-object-state.ts:37]:
```suggestion
export function useSelectedState<T>(
  props: CheckboxObjectStateProps<T>
): CheckboxObjectStateReturn<T> {
  const { compare = (a, b) => a === b, defaultValue = [] } = props

  const [value, setValue] = useState(defaultValue)

  const toggle = (item: T) => {
    if (isSelected(item)) {
      unselect(item)
    } else {
      select(item)
    }
  }

  const isSelected = (candidate: T) => 
    value.some((item) => compare(item, candidate))

  const select = (item: T) => setValue((curr) => [...curr, item]), [value])

  const unselect = (item: T) =>
    setValue((curr) =>
      curr.filter((i) => !compare(i, item))
    )

  return {
    toggle,
    select,
    unselect,
    isSelected,
    value: selectedItems,
    setValue: setSelectedItems,
  }
}
```

**Review by matheusps — COMMENTED** [review]
I think you just need to describe how items are compared, and not a getValue function. Also, the behavior name could be not tied to how it's rendered - that's why `useSelectedState`.
What do you think?

File: packages/admin-ui/src/combobox/combobox-multiple-popover.tsx [hunk 8]
```diff
@@ -3,27 +3,47 @@ import { createComponent, useElement } from '@vtex/admin-ui-react'
...
+            if (isSelected) {
+              // remove
+              props.state.removeSelectedItem(item)
+            } else {
+              // insert
+              props.state.addSelectedItem(item)
+            }
+
+            props.state.setSelectedItem(item)
+          }}
```
[the remark this answers is not in the document]
Comment (anitavincent) [reply packages/admin-ui/src/combobox/combobox-multiple-popover.tsx]: That code has been removed already 🤔

File: packages/admin-ui/src/combobox/combobox-multiple-item.tsx [hunk 9]
```diff
@@ -47,5 +53,6 @@ export function ComboboxMultipleItem(props: ComboboxMultipleItemProps) {
 }
 
 export type ComboboxMultipleItemProps = ComboboxItemProps & {
+  onItemSelect: (isSelected: boolean) => void
```
[the remark this answers is not in the document]
Comment (anitavincent) [reply packages/admin-ui/src/combobox/combobox-multiple-item.tsx]: This code has also been removed already 🤷 I think you're somehow viewing the older commit here

File: packages/admin-ui/src/combobox/combobox-multiple-field.tsx [hunk 10]
```diff
@@ -51,6 +52,15 @@ export const ComboboxMultipleField = createComponent<
...
   } = props
 
+  const {
+    selectedItems,
+    value: inputValue,
+    removeSelectedItem,
+    clearSelected,
+    getOptionValue,
+    renderTag,
+  } = state
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-multiple-field.tsx:62]: I would keep this in the `state` variable. It is easier to read and you don't need to rename things

File: packages/admin-ui/src/combobox/combobox-multiple-item.tsx [hunk 11]
```diff
@@ -17,35 +25,39 @@ export function ComboboxMultipleItem(props: ComboboxMultipleItemProps) {
...
+          // Disable `checked` and `aria-checked` attributes so they don't
+          // conflict with the `aria-selected` attribute.
+          aria-checked={undefined}
+          checked={undefined}
+          value={value}
+          as="div"
+        >
+          {children}
+          <CheckboxCheck checked={checked} />
+        </Checkbox>
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-multiple-item.tsx:53]: You don't need this extra code, right?

File: packages/admin-ui/src/combobox/selected-multiple.state.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import type { Dispatch, SetStateAction } from 'react'
+import { useState } from 'react'
+
+export function useCheckboxObjectState<T>(
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/selected-multiple.state.ts:4]: Maybe `useMultipleSelectionState` ? You don't necessarily need to render a checkbox.

File: packages/admin-ui/src/combobox/combobox-multiple.state.ts [hunk 13]
```diff
@@ -1,54 +1,80 @@
...
 
-  const removeSelected = useCallback(
-    (value: string) => {
-      checkbox.setValue((values: string[]) =>
-        values.filter((currentValue) => currentValue !== value)
-      )
-    },
-    [checkbox]
-  )
+  const addSelectedItem = (newItem: T) => checkbox.select(newItem)
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-multiple.state.ts:42]: I would keep the `select` and unselect `names`. This way you don't need to wrap anything (see the next suggestion).

File: packages/admin-ui/src/combobox/combobox-multiple.state.ts [hunk 14]
```diff
@@ -1,54 +1,80 @@
...
+  const combobox = useComboboxState<T>({
     // VoiceOver has issues with multi-selectable comboboxes where the DOM focus
     // is on the combobox input, so we set `virtualFocus` to `false` to disable
     // this behavior and put DOM focus on the items.
     virtualFocus: false,
     list,
   })
 
-  const checkbox = useCheckboxState({
+  const checkbox = useCheckboxObjectState({
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-multiple.state.ts:32]:
```suggestion
  const { select, unselect } = useCheckboxObjectState({
```
then export it directly :)

File: packages/admin-ui/src/combobox/combobox-multiple-item.tsx [hunk 15]
```diff
@@ -17,35 +24,29 @@ export function ComboboxMultipleItem(props: ComboboxMultipleItemProps) {
...
+
   return (
     <ComboboxItem
-      // All selectable items must have the `aria-selected` attribute set to
-      // `true` or `false`.
-      aria-selected={!!value && checkbox?.value.includes(value)}
+      // All selectable items must have the `aria-selected` attribute
+      aria-selected={checked}
+      {...styleProps}
+      {...onChange}
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-multiple-item.tsx:39]:
```suggestion
      onClick={() => onChange.?(item)}
```

File: packages/admin-ui/src/combobox/combobox-multiple-item.tsx [hunk 16]
```diff
@@ -17,35 +24,29 @@ export function ComboboxMultipleItem(props: ComboboxMultipleItemProps) {
     className: cn(style.itemMultiple),
   }
 
+  // TODO: Check the type error that forbids onClick param
+  const onChange: any = {
+    onClick: () => onChangeCb?.(item),
+  }
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-multiple-item.tsx:30]:
```suggestion
```

**Review by matheusps — APPROVED** [review]

**Pull request merged**